libs/ml/examples/word2vec_SGNS_dataloader/main.cpp

```diff
@@ -0,0 +1,279 @@
+
+  TrainingParams tp;
+
+  ///////////////////////////////////////
+  /// CONVERT TEXT INTO TRAINING DATA ///
+  ///////////////////////////////////////
+
+  std::cout << "Setting up training data...: " << std::endl;
+
+  GraphW2VLoader<DataType> data_loader(tp.window_size, tp.negative_sample_size, tp.freq_thresh,
+                                       tp.max_word_count, tp.train_mode);
+  // set up dataloader
+  /// DATA LOADING ///
+  data_loader.BuildVocab({ReadFile(train_file)}, tp.min_count);
+
+  /////////////////////////////////////////
+  /// SET UP PROPER TRAINING PARAMETERS ///
+  /////////////////////////////////////////
+
+  // calc the true starting learning rate
+  tp.starting_learning_rate =
+      static_cast<DataType>(tp.batch_size) * tp.starting_learning_rate_per_sample;
+  tp.ending_learning_rate =
+      static_cast<DataType>(tp.batch_size) * tp.ending_learning_rate_per_sample;
+  tp.learning_rate_param.starting_learning_rate = tp.starting_learning_rate;
+  tp.learning_rate_param.ending_learning_rate   = tp.ending_learning_rate;
+
+  // calc the compatiable linear lr decay
+  tp.learning_rate_param.linear_decay_rate =
+      static_cast<DataType>(1) /
+      data_loader
+          .EstimatedSampleNumber();  // this decay rate gurantee the lr is reduced to zero by the
+                                     // end of an epoch (despit capping by ending learning rate)
+  std::cout << "data_loader.EstimatedSampleNumber(): " << data_loader.EstimatedSampleNumber()
+            << std::endl;
+
+  ////////////////////////////////
+  /// SETUP MODEL ARCHITECTURE ///
+  ////////////////////////////////
+
+  // set up model architecture
+  std::cout << "building model architecture...: " << std::endl;
+  auto                                g = std::make_shared<fetch::ml::Graph<ArrayType>>();
+  std::pair<std::string, std::string> error_and_skipgram_layer =
+      Model(*g, tp.embedding_size, data_loader.vocab_size());
+  std::string error          = error_and_skipgram_layer.first;
+  std::string skipgram_layer = error_and_skipgram_layer.second;
+
+  /////////////////////////////////
+  /// TRAIN THE WORD EMBEDDINGS ///
+  /////////////////////////////////
+
+  std::cout << "beginning training...: " << std::endl;
+
+  // Initialise Optimiser
+  fetch::ml::optimisers::SGDOptimiser<ArrayType> optimiser(g, {"Input", "Context"}, "Label", error,
+                                                           tp.learning_rate_param);
+
+  // Training loop
+  for (SizeType i{0}; i < tp.training_epochs; i++)
+  {
+    std::cout << "start training for epoch no.: " << i << std::endl;
+    optimiser.Run(data_loader, tp.batch_size);
+    std::cout << std::endl;
+    // Test trained embeddings
+    if (i % tp.test_frequency == 0)
+    {
+      TestEmbeddings(*g, skipgram_layer, data_loader, tp.word0, tp.word1, tp.word2, tp.word3, tp.k);
+      TestEmbeddings(*g, skipgram_layer, data_loader, "france", "france", "paris", "italy", tp.k);
```

Comment:
making two calls to TestEmbeddings once with variables and once with hard-coded values is not great. If you really want to do this you should at least add the hard-coded values as variables to the training params struct and invoke them properly. Then another developer can look at the training params struct and know they exist